src/task.c

```diff
@@ -610,91 +606,45 @@ task_enqueue(task_system *ts,
 void
 task_perform_all(task_system *ts)
 {
-   if (!ts->use_bg_threads) {
-      platform_status rc;
-      do {
-         rc = task_perform_one(ts);
-         debug_assert(SUCCESS(rc) || STATUS_IS_EQ(rc, STATUS_TIMEDOUT));
-      } while (STATUS_IS_NE(rc, STATUS_TIMEDOUT));
-
-   } else {
-      // wait for bg threads to finish running all the queued up tasks.
-      for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
-         task_group *group = &ts->group[type];
-         while (group->current_outstanding_tasks != 0) {
-            platform_sleep(USEC_TO_NSEC(100000)); // 100 msec.
-         }
-      }
-   }
+   platform_status rc;
+   do {
+      rc = task_perform_one(ts);
+      debug_assert(SUCCESS(rc) || STATUS_IS_EQ(rc, STATUS_TIMEDOUT));
+   } while (STATUS_IS_NE(rc, STATUS_TIMEDOUT));
 }
 
 static void
 task_queue_unlock(void *arg)
 {
    task_group *group = (task_group *)arg;
-   platform_assert(!task_system_use_bg_threads(group->ts));
-   platform_mutex_unlock(&group->fg.mutex);
+   task_unlock_task_queue(group);
 }
 
 static inline platform_status
-task_group_perform_one(task_group *group)
+task_group_perform_one(task_group *group, uint64 min_backlog)
 {
-   platform_assert(!task_system_use_bg_threads(group->ts));
    platform_status rc;
    task           *assigned_task = NULL;
-   task_queue     *tq            = &group->tq;
-   if (group->current_outstanding_tasks == 0) {
+   if (group->current_outstanding_tasks <= min_backlog) {
```

Comment:
Is timed out the best status to return here?

Reply:
The only other plausible one I see is NOT_FOUND.  TIMEDOUT is what was used before, and I could see it generalizing in the future if we add time limits to some of these functions.